### PR TITLE
Count lines of code

### LINKED ISSUE #8: Count lines of code?
I think it would be interesting to see things like the distribution of # of LoC over packages in General, or count the total LoC in a registry. To be extra informative, we could do four counts: `src`, `test`, `docs`, and the whole directory (in case there are interesting files in other folders).

I've used `tokei` before and it's really fast and seems quite accurate, and can count Julia code, markdown, etc, and can output JSON or other structured data. So using that, we could either hold counts for common things (like Julia code / markdown / comments) in our own data structure, or just hold onto the JSON blobs in the `Package` and query them later on.

What do you think?

### PR DESCRIPTION
**Pull request by ericphanson**
Closes #8 

WIP because this is based off the #9 branch, is using the system `tokei` (until https://github.com/JuliaPackaging/Yggdrasil/pull/2540 is merged), and still needs tests (and is kind of broken on subdirectory packages, as is some other parts of the package, ref #10).

This adds some LoC counts to the show method:
```julia
julia> analyze_from_registry(joinpath(general_registry(), "B", "BinaryBuilder"))
Package BinaryBuilder:
  * repo: https://github.com/JuliaPackaging/BinaryBuilder.jl.git
  * uuid: 12aac903-9f7c-5d81-afc2-d9565ea332ae
  * is reachable: true
  * has documentation: true
  * has tests: true
  * has continuous integration: true
    * GitHub Actions
    * Azure Pipelines
  * has license(s) in file: MIT
    * filename: LICENSE.md
    * OSI approved: true
  * lines of Julia code in `src`: 4733
  * lines of Julia code in `test`: 1520
 ```
but actually stores much more detailed information. E.g. 

```julia
using AnalyzeRegistry, DataFrames

pkgs = find_packages(; names = ["Flux", "Optim", "MathOptInterface", "Arrow"])

results = analyze_from_registry!(root, pkgs)

df = DataFrame()
for pkg in results
    pkg_df = DataFrame(pkg.lines_of_code)
    pkg_df[!, :name] .= pkg.name
    select!(pkg_df, :name, Not([:name])) # reorder
    append!(df, pkg_df)
end

sort(df, :lines; rev=true)[1:10,:]
```
yields
```julia
10×5 DataFrame
 Row │ name              directory  language  line_type  lines 
     │ String            Symbol     Symbol    Symbol     Int64 
─────┼─────────────────────────────────────────────────────────
   1 │ MathOptInterface  src        Julia     code       40466
   2 │ Arrow             test       JSON      code       38488
   3 │ MathOptInterface  test       Julia     code       12783
   4 │ Optim             src        Julia     code        5811
   5 │ Flux              src        Julia     code        4863
   6 │ Arrow             src        Julia     code        4136
   7 │ MathOptInterface  src        Julia     blanks      3511
   8 │ Optim             test       Julia     code        2820
   9 │ MathOptInterface  src        Julia     comments    2506
  10 │ MathOptInterface  docs       Markdown  comments    2116
```
which I thought was kind of cool.

### CODE REVIEW

**Comment by giordano** [comment]
Side note.  The original version was so simple and fast that my full analysis of the General registry was bound by my (not incredibly fast) Internet connection.  I'm glad now there are more and more things to do, maybe having added multi-threading with `FLoops.jl` will finally make sense :joy:

**Comment by ericphanson** [comment]
I change the table format to match what `tokei` displays:
```julia
shell> tokei .
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Julia                   3          460          390           34           36
 TOML                    5           38           34            0            4
-------------------------------------------------------------------------------
 Markdown                1           68            0           48           20
 |- Julia                1           63           62            0            1
 (Total)                            131           62           48           21
===============================================================================
 Total                   9          566          424           82           60
===============================================================================
```
Now, the table is:
```julia
julia> result.lines_of_code
5-element Vector{NamedTuple{(:directory, :language, :files, :code, :comments, :blanks), Tuple{Symbol, Symbol, Int64, Int64, Int64, Int64}}}:
 (directory = :src, language = :Julia, files = 2, code = 332, comments = 21, blanks = 25)
 (directory = :test, language = :Julia, files = 1, code = 58, comments = 13, blanks = 11)
 (directory = :test, language = :TOML, files = 4, code = 10, comments = 0, blanks = 0)
 (directory = :all_others, language = :Markdown, files = 1, code = 0, comments = 48, blanks = 20)
 (directory = :all_others, language = :TOML, files = 1, code = 24, comments = 0, blanks = 4)
```

Before I couldn't store `files` at all, because I didn't have 1 row per language x directory, I had separate rows for `code`, `comments`, etc. I think this way is more useful and compact, and can always be post-processed into the other format.

**Comment by ericphanson** [comment]
Yeah, `tokei` and `licensecheck` are both pretty fast pieces of code, but I'm sure it will all add up aha. Now we have Package.toml parsing, four calls to `tokei`, a bunch of calls to `licensecheck`, etc..

File: src/AnalyzeRegistry.jl [hunk 1]
```diff
@@ -47,11 +52,13 @@ function Package(name, uuid, repo;
                  license_filename=missing,
                  licenses_found=String[],
                  license_file_percent_covered=missing,
-                 licenses_in_project=String[]
+                 licenses_in_project=String[],
+                 lines_of_code=Vector{LoCTableEltype}()
```
Comment (giordano) [inline src/AnalyzeRegistry.jl:56]:
```suggestion
                 lines_of_code=Vector{LoCTableEltype}(),
```

File: src/AnalyzeRegistry.jl [hunk 2]
```diff
@@ -113,6 +120,10 @@ function Base.show(io::IO, p::Package)
         else
             body *= "  * has continuous integration: false\n"
         end
+        if !isempty(p.lines_of_code)
+          body *= "  * lines of Julia code in `src`: $(count_julia_loc(p.lines_of_code, :src))\n"
+          body *= "  * lines of Julia code in `test`: $(count_julia_loc(p.lines_of_code, :test))\n"
+        end
```
Comment (giordano) [inline src/AnalyzeRegistry.jl:126]: Also here, I think the number of lines is more juicy than CI or documentation, I would move these lines above :slightly_smiling_face:

Comment (ericphanson) [reply src/AnalyzeRegistry.jl]: sounds good! how does this look?
```julia
julia> analyze(pkgdir(AnalyzeRegistry))
Package AnalyzeRegistry:
  * repo: 
  * uuid: e713c705-17e4-4cec-abe0-95bf5bf3e10c
  * is reachable: true
  * lines of Julia code in `src`: 326
  * lines of Julia code in `test`: 58
  * has license(s) in file: MIT
    * filename: LICENSE
    * OSI approved: true
  * has documentation: false
  * has tests: true
  * has continuous integration: true
    * GitHub Actions
```

**Comment by ericphanson** [comment]
I reworked the lines of code table again; now I just call `tokei` once at the toplevel, and then filter/group the results myself. That seems a little cleaner (no more messing with exclude flags) and now I look at results within all top-level files/directories:

```julia
julia> using AnalyzeRegistry, DataFrames

julia> result = analyze(pkgdir(AnalyzeRegistry));

julia> DataFrame(result.lines_of_code)
6×7 DataFrame
 Row │ directory     language  sublanguage  files  code   comments  blanks 
     │ String        Symbol    Union…       Int64  Int64  Int64     Int64  
─────┼─────────────────────────────────────────────────────────────────────
   1 │ src           Julia                      2    327        26      25
   2 │ README.md     Markdown  Julia            1     76         0       3
   3 │ test          Julia                      1     65        13      11
   4 │ Project.toml  Toml                       1     25         0       4
   5 │ test          Toml                       4     10         0       0
   6 │ README.md     Markdown                   1      0        54      23
```

I also added a `sublanguage` column because `tokei` can even identify Julia code within markdown code, so we can see how much of the README are examples.

I'm not sure about calling the first column "directory", since it's not only directories. I thought about grouping all toplevel files into an "other_files" fake directory, but I thought it's actually kind of nice to see README vs other things; I figure if a file is at top-level, maybe it's important.

**Comment by ericphanson** [comment]
Tweaked the sorting so that directories are grouped together:
```julia
julia> DataFrame(result.lines_of_code)
6×7 DataFrame
 Row │ directory     language  sublanguage  files  code   comments  blanks 
     │ String        Symbol    Union…       Int64  Int64  Int64     Int64  
─────┼─────────────────────────────────────────────────────────────────────
   1 │ src           Julia                      2    334        26      26
   2 │ README.md     Markdown                   1      0        54      23
   3 │ README.md     Markdown  Julia            1     76         0       3
   4 │ test          Julia                      1     65        13      11
   5 │ test          Toml                       4     10         0       0
   6 │ Project.toml  Toml                       1     25         0       4
```
It's a little bit more expensive to do this sorting and all, but it should take milliseconds, so for 4k packages it should still be just a couple of seconds more 🙂 .

File: src/AnalyzeRegistry.jl (hunk 2 above)
Comment (giordano) [reply src/AnalyzeRegistry.jl]: Looks good!

**Comment by giordano** [comment]
Sorry for bringing this point at this time, but I'm wondering if [`linguist`](https://github.com/github/linguist/) is a better option because that's what GitHub uses, and should honour `.gitattributes` files.  On the other hand, packaging `linguist` should be much more complicated than Tokei, so maybe we can stick to this one :smile:

**Comment by ericphanson** [comment]
That seems like an interesting option, but I’m not sure it gives us as much info as tokei. It seems like their main concern is not misclassifying the repo language and getting a percent score for each language, so they do things like exclude docs. But for us, we know it’s a Julia package, so that’s not so important and it’s nice to get stats for things like the docs.

**Comment by ericphanson** [comment]
Ok with the new Tokei_jll release, this is good to go from my point of view 🙂. (Should be squash-merged as opposed to merge-committed IMO, since there's a lot of random commits).

**Review by giordano — APPROVED** [review]

**Pull request merged**